### ISSUE TITLE
Optimization for beamlaser and lightningcannon units enhancement

### ISSUE THREAD

**Issue by 6AKU66** [beyond-all-reason/Beyond-All-Reason#2043]
Beamlaser and lightningcannon type of weapon is pretty heavy.
Most obvious way for optimization for beamlaser and lightningcannon is apply damage only once in 2-3 ticks. This method have some problem: Laserbeam/lightningbeam ray is flickering.
And could be more problems with it, need more testing.

https://www.youtube.com/watch?v=eKO0IO8J8zw (Hit every 1 tick)
https://www.youtube.com/watch?v=vhzk0N9XxKU (Hit every 3 tick)
For (Hit every 3 tick) i applied this setting for armbanth unit:
beamburst = true,
beamdecay = 100,
beamtime = 0.07,
burst = 300,
burstrate = 0.0999,

**Comment by sprunk** [comment]
> Most obvious way for optimization for beamlaser and lightningcannon is apply damage only once in 2-3 ticks.

Lightning cannon already applies its damage in a single tick and then leaves a graphical effect for N frames that does not do any damage or collision checks, just that BAR specifically configures it to instead spawn N lightnings instead with a lifetime of 1 frame. This doesn't necessarily mean it can be solved gameside because the lingering lightning is essentially static (ha) and thus somewhat ugly (compare with ZK which so far still uses the ugly lightning, `cloakassault` for a unit name example).

For beamlaser the "damage and raytrace each tick" behaviour is important because mechanics may depend on it:
* a game could have some sort of buffing beam (think Tukohama in PoE, or maybe a hypothetical blockable version of Wisp Tether in DotA) implemented to grant something for 1 frame. Then any gaps would ruin the effect despite it seemingly being maintained. Of course you can just change it to grant the thing for longer but this complicates things, especially if the mechanic can be stacked multiple times, and generally reduces the ceiling of how polished things can be.
* in particular for the above, I believe TAP currently does this. Their resources work a bit like in starcraft where you've got welders beaming rocks for resources, which interacts with how the resource drain system works every frame. Larger chunks means that things with discrete usage now get the opportunity to drain it (this is how laser weapons get to work in BAR even if you're stalling) which can be a large design change.
* one of ZK mods, and I think some of the LD games, have gravity beams. As far as I can tell the engine has something akin to static friction (I am unsure if it's actually implemented as such or just there's some magic impulse threshold, but that doesn't really matter) which chunky impulse would overcome more easily, and impulse is also pretty opaque as a mechanic as far as replicating it in Lua.
* ZK Starlight (note, `mahlazer`, not the same as BAR `armmanni`) making smooth trenches via cratering probably also counts. Having to replicate cratering in Lua sounds terrible.
* damage/health interactions. If a unit dies in N frames, then making the beam deal damage once per N-1 or N+1 chunks can affect the relationship. This and the above concern could apply to any powerful sweeping laser (think Supcom Monkeylord).
* it is not necessarily obvious how this works for generalized values. For example what if your beam lasts 8 frames and you set it to deal damage every 5?

**Comment by 6AKU66** [comment]
@sprunk Lightning cannon rays itself not staying though.
https://youtu.be/OfLrNsLFJC8

**Comment by sprunk** [comment]
set beamTTL=10

**Comment by 6AKU66** [comment]
Ok, it's working. The only problem that it's looking too slo-mo'ish because lightning ray don't changing shape.
https://youtu.be/1L3tdh_wEXo

**Comment by sprunk** [comment]
Yes, that's what I meant when I said it's static and ugly. I don't think this can be fixed gameside just via weapon def tags.
